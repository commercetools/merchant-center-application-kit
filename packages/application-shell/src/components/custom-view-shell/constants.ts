// TODO: This file is duplicated (also exists in application-components custom views components)
export const CUSTOM_VIEWS_EVENTS_NAMES = {
  CUSTOM_VIEW_BOOTSTRAP: 'custom-view-bootstrap',
  CUSTOM_VIEW_INITIALIZATION: 'custom-view-initialization',
};

export const CUSTOM_VIEWS_EVENTS_META = {
  SOURCE: 'mc-host-application',
  DESTINATION_PREFIX: 'custom-view-',
};
