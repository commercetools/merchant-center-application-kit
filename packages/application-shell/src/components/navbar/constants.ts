export const NAVBAR = {
  expanderHeight: '50px',
  expanderSize: '40px',
  iconSize: '24px',
  iconSizeHover: '28px',
  itemSize: '48px',
  itemHeight: '56px',
  sublistIndentationWhenCollapsed: '72px',
  sublistIndentationWhenExpanded: '248px',
  sublistWidth: '253px',
  leftNavigationTransition: 'all 150ms cubic-bezier(1, 0, 0.58, 1)',
  widthLeftNavigation: '80px',
  widthLeftNavigationWhenExpanded: '256px',
} as const;
