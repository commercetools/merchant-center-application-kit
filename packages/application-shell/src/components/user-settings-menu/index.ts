export { default } from './user-settings-menu';
