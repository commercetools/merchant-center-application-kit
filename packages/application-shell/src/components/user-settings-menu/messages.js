import { defineMessages } from 'react-intl';

export default defineMessages({
  userProfile: {
    id: 'UserSettingsMenu.userProfile',
    description: 'The label for user profile option',
    defaultMessage: 'My Profile',
  },
  manageOrganizations: {
    id: 'UserSettingsMenu.manageOrganizations',
    description: 'The label for manage organizations option',
    defaultMessage: 'Manage Organizations & Teams',
  },
  manageProjects: {
    id: 'UserSettingsMenu.manageProjects',
    description: 'The label for manage projects option',
    defaultMessage: 'Manage Projects',
  },
  manageTeams: {
    id: 'UserSettingsMenu.manageTeams',
    description: 'The label for manage teams option',
    defaultMessage: 'Manage Teams',
  },
  support: {
    id: 'UserSettingsMenu.support',
    description: 'The label for Support',
    defaultMessage: 'Support',
  },
  logout: {
    id: 'UserSettingsMenu.logout',
    description: 'The label for logout option',
    defaultMessage: 'Logout',
  },
  privacyPolicy: {
    id: 'UserSettingsMenu.privacyPolicy',
    description: 'The label for privacy policy option',
    defaultMessage: 'Privacy Policy',
  },
});
