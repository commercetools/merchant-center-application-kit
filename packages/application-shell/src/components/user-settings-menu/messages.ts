import { defineMessages } from 'react-intl';

export default defineMessages({
  support: {
    id: 'UserSettingsMenu.support',
    description: 'The label for Support',
    defaultMessage: 'Support',
  },
  logout: {
    id: 'UserSettingsMenu.logout',
    description: 'The label for logout option',
    defaultMessage: 'Logout',
  },
  privacyPolicy: {
    id: 'UserSettingsMenu.privacyPolicy',
    description: 'The label for privacy policy option',
    defaultMessage: 'Privacy Policy',
  },
  openMenuLabel: {
    id: 'UserSettingsMenu.openMenuLabel',
    description: 'The label when menu is closed and would be opened by a click',
    defaultMessage: 'Open user settings menu',
  },
  closeMenuLabel: {
    id: 'UserSettingsMenu.closeMenuLabel',
    description: 'The label when menu is open and would be closed by a click',
    defaultMessage: 'Close user settings menu',
  },
});
