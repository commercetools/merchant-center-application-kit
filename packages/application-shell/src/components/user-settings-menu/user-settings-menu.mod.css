:root {
  --menu-width: 315px;
  --menu-position-top: 40px;
  --menu-position-right: 14px;
  --menu-offset-right: 28px;
}

.settings-container {
  cursor: pointer;
  border: none;
  padding: 0;
  display: flex;
  background: transparent;
}

.menu {
  position: absolute;
  border: 1px var(--color-green-40) solid;
  box-shadow: var(--shadow-7-first), var(--shadow-7-second);
  width: var(--menu-width);
  right: var(--menu-position-right);
  top: var(--menu-position-top);
  padding: var(--spacing-4);
  overflow: hidden;
}

.menu::before {
  content: '';
  position: absolute;
  right: 30px;
  top: -6px;
  border-style: solid;
  border-width: 0 6px 6px;
  z-index: 1;
}

.menu::after {
  content: '';
  position: absolute;
  right: var(--menu-offset-right);
  top: calc(var(--spacing-8) * -1);
  border-style: solid;
  border-width: 0 var(--spacing-8) var(--spacing-8);
  z-index: 0;
}

.item {
  width: 100%;
  cursor: pointer;
  color: var(--color-black);
}

.item:hover {
  background-color: var(--color-gray-90);
}

.item-divider-account-section {
  border-bottom: 1px solid var(--color-gray);
}

.email-divider {
  margin-bottom: var(--spacing-4);
}
