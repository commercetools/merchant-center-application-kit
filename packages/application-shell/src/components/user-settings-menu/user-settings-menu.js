import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import classnames from 'classnames';
import Downshift from 'downshift';
import { ToggleFeature } from '@flopflip/react-broadcast';
import {
  withMouseOverState,
  CaretDownIcon,
  Text,
  Spacings,
  Avatar,
} from '@commercetools-frontend/ui-kit';
import { LOGOUT_REASONS } from '@commercetools-frontend/constants';
import Card from '../../from-core/card';
import { MCSupportFormURL } from '../../constants';
import { PROJECTS_LIST, ORGANIZATIONS_LIST } from '../../feature-toggles';
import styles from './user-settings-menu.mod.css';
import messages from './messages';

export const UserAvatar = props => (
  <div onMouseOver={props.handleMouseOver} onMouseOut={props.handleMouseOut}>
    <Spacings.Inline alignItems="center">
      <Avatar
        gravatarHash={props.gravatarHash}
        firstName={props.firstName}
        lastName={props.lastName}
        isHighlighted={props.isMouseOver}
      />
      <CaretDownIcon
        size="small"
        theme={props.isMouseOver ? 'grey' : 'black'}
      />
    </Spacings.Inline>
  </div>
);
UserAvatar.displayName = 'UserAvatar';
UserAvatar.propTypes = {
  gravatarHash: PropTypes.string.isRequired,
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  email: PropTypes.string,
  // Injected
  handleMouseOver: PropTypes.func.isRequired,
  handleMouseOut: PropTypes.func.isRequired,
  isMouseOver: PropTypes.bool.isRequired,
};
const UserAvatarWithHoverState = withMouseOverState(UserAvatar);

function stateReducer(state, changes) {
  switch (changes.type) {
    // So in case the user wants to navigate with the tab button
    // we need to make sure that the menu does not close
    case Downshift.stateChangeTypes.blurButton:
      return {
        ...changes,
        isOpen: true,
      };
    default:
      return changes;
  }
}

export class UserSettingsMenu extends React.PureComponent {
  static displayName = 'UserSettingsMenu';

  static propTypes = {
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    email: PropTypes.string,
    gravatarHash: PropTypes.string.isRequired,
    isAccountPath: PropTypes.bool.isRequired,
  };

  render() {
    return (
      <div data-test="user-settings-menu">
        <Downshift stateReducer={stateReducer}>
          {({ isOpen, toggleMenu, getToggleButtonProps, getMenuProps }) => (
            <div>
              <button
                className={styles['settings-container']}
                {...getToggleButtonProps()}
              >
                <UserAvatarWithHoverState
                  firstName={this.props.firstName}
                  lastName={this.props.lastName}
                  gravatarHash={this.props.gravatarHash}
                />
              </button>
              {isOpen && (
                <Card className={styles.menu}>
                  <div {...getMenuProps()}>
                    <div
                      className={classnames({
                        [styles['item-divider-account-section']]: this.props
                          .isAccountPath,
                      })}
                    >
                      <Spacings.Inset scale="xs">
                        <Spacings.Inline scale="xs" alignItems="center">
                          <Avatar
                            firstName={this.props.firstName}
                            lastName={this.props.lastName}
                            gravatarHash={this.props.gravatarHash}
                          />
                          <div
                            className={classnames({
                              [styles['email-divider']]: this.props
                                .isAccountPath,
                            })}
                          >
                            <Text.Body isBold>
                              {[this.props.firstName, this.props.lastName]
                                .join(' ')
                                .trim()}
                            </Text.Body>
                            <Text.Body truncate>{this.props.email}</Text.Body>
                          </div>
                        </Spacings.Inline>
                      </Spacings.Inset>
                    </div>
                    {!this.props.isAccountPath && (
                      <React.Fragment>
                        <Link to="/account/profile" onClick={toggleMenu}>
                          <div className={styles.item}>
                            <Spacings.Inset scale="s">
                              <FormattedMessage {...messages.userProfile} />
                            </Spacings.Inset>
                          </div>
                        </Link>
                        <ToggleFeature flag={ORGANIZATIONS_LIST}>
                          <Link
                            to="/account/organizations"
                            onClick={toggleMenu}
                          >
                            <div className={styles.item}>
                              <Spacings.Inset scale="s">
                                <FormattedMessage
                                  {...messages.manageOrganizations}
                                />
                              </Spacings.Inset>
                            </div>
                          </Link>
                        </ToggleFeature>
                        <ToggleFeature flag={PROJECTS_LIST}>
                          <Link to="/account/projects" onClick={toggleMenu}>
                            <div
                              className={classnames(
                                styles.item,
                                styles['item-divider-account-section']
                              )}
                            >
                              <Spacings.Inset scale="s">
                                <FormattedMessage
                                  {...messages.manageProjects}
                                />
                              </Spacings.Inset>
                            </div>
                          </Link>
                        </ToggleFeature>
                      </React.Fragment>
                    )}
                    <a
                      href={`https://commercetools.com/privacy`}
                      target="_blank"
                      onClick={toggleMenu}
                    >
                      <div className={styles.item}>
                        <Spacings.Inset scale="s">
                          <FormattedMessage {...messages.privacyPolicy} />
                        </Spacings.Inset>
                      </div>
                    </a>
                    <a
                      href={MCSupportFormURL}
                      rel="noopener noreferrer"
                      target="_blank"
                      data-track-component="Navigation-Support-links"
                      data-track-event="click"
                      data-track-label="support_textlink"
                      onClick={toggleMenu}
                    >
                      <div
                        className={classnames(
                          styles.item,
                          styles['item-divider-account-section']
                        )}
                      >
                        <Spacings.Inset scale="s">
                          <FormattedMessage {...messages.support} />
                        </Spacings.Inset>
                      </div>
                    </a>
                    <a
                      // NOTE: we want to redirect to a new page so that the
                      // server can remove things like cookie for access token.
                      href={`/logout?reason=${LOGOUT_REASONS.USER}`}
                      data-test="logout-button"
                    >
                      <div className={styles.item} tabIndex="0">
                        <Spacings.Inset scale="s">
                          <FormattedMessage {...messages.logout} />
                        </Spacings.Inset>
                      </div>
                    </a>
                  </div>
                </Card>
              )}
            </div>
          )}
        </Downshift>
      </div>
    );
  }
}

export default UserSettingsMenu;
