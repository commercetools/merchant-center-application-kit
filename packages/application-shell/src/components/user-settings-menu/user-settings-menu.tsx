import type { ControllerStateAndHelpers, DownshiftProps } from 'downshift';
import type { TUser } from '../../types/generated/mc';
import type {
  TApplicationsMenu,
  TFetchApplicationsMenuQuery,
} from '../../types/generated/proxy';

import React from 'react';
import { Link } from 'react-router-dom';
import { useIntl, FormattedMessage } from 'react-intl';
import { css } from '@emotion/react';
import styled from '@emotion/styled';
import Downshift from 'downshift';
import { ToggleFeature } from '@flopflip/react-broadcast';
import Avatar from '@commercetools-uikit/avatar';
import Spacings from '@commercetools-uikit/spacings';
import Text from '@commercetools-uikit/text';
import { CaretDownIcon } from '@commercetools-uikit/icons';
import { customProperties } from '@commercetools-uikit/design-system';
import {
  LOGOUT_REASONS,
  NO_VALUE_FALLBACK,
  SUPPORT_PORTAL_URL,
} from '@commercetools-frontend/constants';
import { useApplicationContext } from '@commercetools-frontend/application-shell-connectors';
import { reportErrorToSentry } from '@commercetools-frontend/sentry';
import useApplicationsMenu from '../../hooks/use-applications-menu';
import messages from './messages';

type Props = Pick<
  TUser,
  'language' | 'firstName' | 'lastName' | 'email' | 'gravatarHash'
> & {
  DEV_ONLY__loadAppbarMenuConfig?: () => Promise<TApplicationsMenu['appBar']>;
};
type MenuBodyProps = Props & {
  downshiftProps: ControllerStateAndHelpers<{}>;
};
type OptionalFeatureToggleProps = {
  featureToggle?: string;
  children: React.ReactNode;
};
type MenuItemProps = {
  hasDivider?: boolean;
};
type MenuConfig = TFetchApplicationsMenuQuery['applicationsMenu']['appBar'][0];

const UserAvatar = (
  props: Pick<Props, 'firstName' | 'lastName' | 'gravatarHash'>
) => {
  const [isMouseOver, setIsMouseOver] = React.useState(false);
  const handleMouseOver = React.useCallback(() => {
    setIsMouseOver(true);
  }, []);
  const handleMouseOut = React.useCallback(() => {
    setIsMouseOver(false);
  }, []);
  return (
    <div onMouseOver={handleMouseOver} onMouseOut={handleMouseOut}>
      <Spacings.Inline alignItems="center">
        <Avatar
          gravatarHash={props.gravatarHash}
          firstName={props.firstName}
          lastName={props.lastName}
          isHighlighted={isMouseOver}
        />
        <CaretDownIcon
          size="small"
          color={isMouseOver ? 'neutral60' : 'solid'}
        />
      </Spacings.Inline>
    </div>
  );
};
UserAvatar.displayName = 'UserAvatar';

function getIsFocusedElementInMenu() {
  return Boolean(
    document.activeElement?.getAttribute('data-user-settings-menu')
  );
}

const stateReducer: DownshiftProps<{}>['stateReducer'] = (state, changes) => {
  switch (changes.type) {
    // So in case the user wants to navigate with the tab button
    // we need to make sure that the menu does not close
    case Downshift.stateChangeTypes.blurButton:
      return {
        ...changes,
        isOpen: (state.isOpen && getIsFocusedElementInMenu()) ?? false,
      };
    default:
      return changes;
  }
};

const OptionalFeatureToggle = (props: OptionalFeatureToggleProps) => {
  if (props.featureToggle) {
    return (
      <ToggleFeature flag={props.featureToggle}>{props.children}</ToggleFeature>
    );
  }
  return <>{props.children}</>;
};

const renderLabel = (
  menu: MenuConfig,
  applicationLanguage: Props['language']
) => {
  const localizedLabel = menu.labelAllLocales.find((loc) =>
    applicationLanguage.startsWith(loc.locale)
  );
  if (localizedLabel) return localizedLabel.value;
  return NO_VALUE_FALLBACK;
};

const MenuItem = styled.div<MenuItemProps>`
  width: 100%;
  cursor: pointer;
  color: ${customProperties.colorSolid};

  :hover {
    background-color: ${customProperties.colorNeutral90};
  }

  ${(props) =>
    props.hasDivider === true
      ? css`
          border-bottom: 1px solid ${customProperties.colorNeutral};
        `
      : ''};
`;

const getUserSettingsMenuItemLinkStyles = () => css`
  display: block;
`;

const UserSettingsMenuBody = (props: MenuBodyProps) => {
  // Focus on a menu item when it's opened through keyboard
  const menuElementRef = React.useRef<HTMLAnchorElement>(null);
  React.useEffect(() => {
    menuElementRef.current?.focus();
  }, []);

  const environment = useApplicationContext((context) => context.environment);
  const applicationsAppBarMenu = useApplicationsMenu<'appBar'>('appBar', {
    queryOptions: {
      onError: reportErrorToSentry,
    },
    environment,
    loadMenuConfig: props.DEV_ONLY__loadAppbarMenuConfig,
  });

  return (
    <div
      onKeyDown={(event) => {
        if (
          event.key === 'Esc' ||
          (event.key === 'Tab' && !getIsFocusedElementInMenu())
        ) {
          props.downshiftProps.closeMenu();
        }
      }}
      css={css`
        position: absolute;
        background: ${customProperties.colorSurface};
        border: 1px ${customProperties.colorPrimary40} solid;
        border-radius: ${customProperties.borderRadius6};
        box-shadow: ${customProperties.shadow7};
        width: 315px;
        right: 14px;
        top: 40px;
        padding: ${customProperties.spacingXs};
        overflow: hidden;
      `}
    >
      <div {...props.downshiftProps.getMenuProps()}>
        <Spacings.Inset scale="xs">
          <Spacings.Inline scale="xs" alignItems="center">
            <Avatar
              firstName={props.firstName}
              lastName={props.lastName}
              gravatarHash={props.gravatarHash}
            />
            <div>
              <Text.Body isBold>
                {[props.firstName, props.lastName].join(' ').trim()}
              </Text.Body>
              <Text.Body truncate>{props.email}</Text.Body>
            </div>
          </Spacings.Inline>
        </Spacings.Inset>
        {applicationsAppBarMenu?.map((menu) => (
          <OptionalFeatureToggle
            key={menu.key}
            featureToggle={menu.featureToggle}
          >
            <Link
              css={getUserSettingsMenuItemLinkStyles()}
              to={`/account/${menu.uriPath}`}
              onClick={() => props.downshiftProps.toggleMenu()}
              data-user-settings-menu
              ref={menuElementRef}
            >
              <MenuItem>
                <Spacings.Inset scale="s">
                  {renderLabel(menu, props.language)}
                </Spacings.Inset>
              </MenuItem>
            </Link>
          </OptionalFeatureToggle>
        ))}
        <MenuItem hasDivider={true} />
        <a
          css={getUserSettingsMenuItemLinkStyles()}
          href={`https://commercetools.com/privacy#suppliers`}
          target="_blank"
          rel="noopener noreferrer"
          onClick={() => props.downshiftProps.toggleMenu()}
          data-user-settings-menu
          ref={!applicationsAppBarMenu ? menuElementRef : undefined}
        >
          <MenuItem>
            <Spacings.Inset scale="s">
              <FormattedMessage {...messages.privacyPolicy} />
            </Spacings.Inset>
          </MenuItem>
        </a>
        <a
          css={getUserSettingsMenuItemLinkStyles()}
          href={SUPPORT_PORTAL_URL}
          rel="noopener noreferrer"
          target="_blank"
          data-track-component="Navigation-Support-links"
          data-track-event="click"
          data-track-label="support_textlink"
          onClick={() => props.downshiftProps.toggleMenu()}
          data-user-settings-menu
        >
          <MenuItem>
            <Spacings.Inset scale="s">
              <FormattedMessage {...messages.support} />
            </Spacings.Inset>
          </MenuItem>
        </a>
        <MenuItem hasDivider={true} />
        <a
          css={getUserSettingsMenuItemLinkStyles()}
          // NOTE: we want to redirect to a new page so that the
          // server can remove things like cookie for access token.
          href={`/logout?reason=${LOGOUT_REASONS.USER}`}
          data-test="logout-button"
          data-user-settings-menu
        >
          <MenuItem>
            <Spacings.Inset scale="s">
              <FormattedMessage {...messages.logout} />
            </Spacings.Inset>
          </MenuItem>
          <div tabIndex={0} onFocus={() => props.downshiftProps.closeMenu()} />
        </a>
      </div>
    </div>
  );
};
UserSettingsMenuBody.displayName = 'UserSettingsMenuBody';

const UserSettingsMenu = (props: Props) => {
  const intl = useIntl();

  return (
    <div data-test="user-settings-menu">
      <Downshift stateReducer={stateReducer}>
        {(downshiftProps) => (
          <div>
            <button
              role="user-menu-toggle"
              css={css`
                cursor: pointer;
                border: none;
                padding: 0;
                display: flex;
                background: transparent;
              `}
              {...downshiftProps.getToggleButtonProps({
                'aria-label': intl.formatMessage(
                  downshiftProps.isOpen
                    ? messages.closeMenuLabel
                    : messages.openMenuLabel
                ),
              })}
            >
              <UserAvatar
                firstName={props.firstName}
                lastName={props.lastName}
                gravatarHash={props.gravatarHash}
              />
            </button>
            {downshiftProps.isOpen && (
              <UserSettingsMenuBody
                {...props}
                downshiftProps={downshiftProps}
              />
            )}
          </div>
        )}
      </Downshift>
    </div>
  );
};
UserSettingsMenu.displayName = 'UserSettingsMenu';

export default UserSettingsMenu;

// For testing
export { UserAvatar, UserSettingsMenuBody };
