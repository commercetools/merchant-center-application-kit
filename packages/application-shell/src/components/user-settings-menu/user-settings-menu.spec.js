import React from 'react';
import { shallow } from 'enzyme';
import { CaretDownIcon, Avatar } from '@commercetools-frontend/ui-kit';
import Downshift from 'downshift';
import { MCSupportFormURL } from '../../constants';
import { UserSettingsMenu, UserAvatar } from './user-settings-menu';

const createTestProps = props => ({
  firstName: 'John Test',
  lastName: 'Doe',
  email: 'john@doe.com',
  gravatarHash: '20c9c1b252b46ab49d6f7a4cee9c3e68',
  ...props,
});

const createDownshiftProps = props => ({
  isOpen: false,
  toggleMenu: jest.fn(),
  getToggleButtonProps: jest.fn(),
  getMenuProps: jest.fn(),
  ...props,
});

describe('rendering', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<UserSettingsMenu {...props} />);
  });

  it('should render <Downshift> wrapper', () => {
    expect(wrapper).toRender(Downshift);
  });

  describe('menu', () => {
    let downshiftProps;
    let menuStateContainerRenderWrapper;
    beforeEach(() => {
      downshiftProps = createDownshiftProps();
      menuStateContainerRenderWrapper = shallow(
        wrapper.find(Downshift).prop('children')(downshiftProps)
      );
    });
    it('should render button', () => {
      expect(menuStateContainerRenderWrapper).toRender('button');
    });
    it('should render WithMouseOverState(UserAvatar)', () => {
      expect(menuStateContainerRenderWrapper).toRender(
        'WithMouseOverState(UserAvatar)'
      );
    });
    describe('when menu is open', () => {
      beforeEach(() => {
        downshiftProps = createDownshiftProps({ isOpen: true });
        menuStateContainerRenderWrapper = shallow(
          wrapper.find(Downshift).prop('children')(downshiftProps)
        );
      });
      it('should render matching tree', () => {
        expect(menuStateContainerRenderWrapper).toMatchSnapshot();
      });
      it('should render link to "/account/profile"', () => {
        expect(menuStateContainerRenderWrapper).toRender({
          to: '/account/profile',
        });
      });
      it('should render link to "/logout"', () => {
        expect(menuStateContainerRenderWrapper).toRender({
          href: '/logout?reason=user',
        });
      });
      it('should render link to "MCSupportFormURL', () => {
        expect(menuStateContainerRenderWrapper).toRender({
          href: MCSupportFormURL,
        });
      });
    });
    describe('when menu is closed', () => {
      beforeEach(() => {
        downshiftProps = createDownshiftProps();
        menuStateContainerRenderWrapper = shallow(
          wrapper.find(Downshift).prop('children')(downshiftProps)
        );
      });
      it('should not render <Card>', () => {
        expect(menuStateContainerRenderWrapper).not.toRender('Card');
      });
    });
  });

  describe('<UserAvatar>', () => {
    const createUserAvatarProps = custom => ({
      ...createTestProps(),
      handleMouseOver: jest.fn(),
      handleMouseOut: jest.fn(),
      isMouseOver: false,
      ...custom,
    });
    beforeEach(() => {
      props = createUserAvatarProps();
      wrapper = shallow(<UserAvatar {...props} />);
    });

    it('should render `CaretDownIcon`', () => {
      expect(wrapper).toRender(CaretDownIcon);
    });

    it('should render `Avatar` component', () => {
      expect(wrapper).toRender(Avatar);
    });

    it('should pass prop firstName to `Avatar` component', () => {
      expect(wrapper.find(Avatar)).toHaveProp('firstName', 'John Test');
    });

    it('should pass prop lastName to `Avatar` component', () => {
      expect(wrapper.find(Avatar)).toHaveProp('lastName', 'Doe');
    });

    it('should pass prop `gravatarHash` to `Avatar` component', () => {
      expect(wrapper.find(Avatar)).toHaveProp(
        'gravatarHash',
        '20c9c1b252b46ab49d6f7a4cee9c3e68'
      );
    });
  });
});
