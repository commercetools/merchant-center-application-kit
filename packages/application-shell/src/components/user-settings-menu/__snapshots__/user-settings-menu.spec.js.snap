// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering menu when menu is open should render matching tree 1`] = `
<div>
  <div
    className="settings-container"
    onClick={[MockFunction]}
  >
    <WithMouseOverState(UserAvatar)
      firstName="John Test"
      gravatarHash="20c9c1b252b46ab49d6f7a4cee9c3e68"
      lastName="Doe"
    />
  </div>
  <Card
    className="menu"
    theme="light"
    type="raised"
  >
    <Inset
      scale="s"
    >
      <TextDetail
        tone="secondary"
      >
        John Test Doe
      </TextDetail>
    </Inset>
    <Link
      onClick={[MockFunction]}
      replace={false}
      to="/account/profile"
    >
      <div
        className="item"
      >
        <Inset
          scale="s"
        >
          <FormattedMessage
            defaultMessage="My Profile"
            description="The label for user profile option"
            id="UserSettingsMenu.userProfile"
            values={Object {}}
          />
        </Inset>
      </div>
    </Link>
    <ToggleFeature
      flag="organizationsList"
    >
      <Link
        onClick={[MockFunction]}
        replace={false}
        to="/account/organizations"
      >
        <div
          className="item"
        >
          <Inset
            scale="s"
          >
            <FormattedMessage
              defaultMessage="Manage Organizations"
              description="The label for manage organizations option"
              id="UserSettingsMenu.manageOrganizations"
              values={Object {}}
            />
          </Inset>
        </div>
      </Link>
    </ToggleFeature>
    <ToggleFeature
      flag="projectsList"
    >
      <Link
        onClick={[MockFunction]}
        replace={false}
        to="/account/projects"
      >
        <div
          className="item item-divider-account-section"
        >
          <Inset
            scale="s"
          >
            <FormattedMessage
              defaultMessage="Manage Projects"
              description="The label for manage projects option"
              id="UserSettingsMenu.manageProjects"
              values={Object {}}
            />
          </Inset>
        </div>
      </Link>
    </ToggleFeature>
    <a
      href="https://commercetools.com/privacy"
      target="_blank"
    >
      <div
        className="item"
      >
        <Inset
          scale="s"
        >
          <FormattedMessage
            defaultMessage="Privacy Policy"
            description="The label for privacy policy option"
            id="UserSettingsMenu.privacyPolicy"
            values={Object {}}
          />
        </Inset>
      </div>
    </a>
    <a
      data-track-component="Navigation-Support-links"
      data-track-event="click"
      data-track-label="support_textlink"
      href="https://jira.commercetools.com/servicedesk/customer/portal/1/create/167"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        className="item"
      >
        <Inset
          scale="s"
        >
          <FormattedMessage
            defaultMessage="Support"
            description="The label for Support"
            id="UserSettingsMenu.support"
            values={Object {}}
          />
        </Inset>
      </div>
    </a>
    <a
      data-test="logout-button"
      href="/logout?reason=user"
    >
      <div
        className="item"
      >
        <Inset
          scale="s"
        >
          <FormattedMessage
            defaultMessage="Logout"
            description="The label for logout option"
            id="UserSettingsMenu.logout"
            values={Object {}}
          />
        </Inset>
      </div>
    </a>
  </Card>
</div>
`;
